File: data_preprocessing_template.py
# -*- coding: utf-8 -*-
"""
Created on Tue Dec  4 01:30:59 2018

@author: hazzaldo
"""

# Importing the libraries:
# =========================
# numpy is a library that contains mathematical tools. 
# We will need this library for any mathematical operations. We created 'np' 
# as the shortcut so we don't have to reference the long name of the library.
import numpy as np
#From the matplotlib library we're importing the sub-library pyplot.
#This library help us plot charts. 
import matplotlib.pyplot as plt
#pandas library is the best library for importing and managing datasets.
import pandas as pd


# Importing the dataset:
# =========================
dataset = pd.read_csv('Data.csv')

# Create the Matrix Of Features (independant variables):
#Left of the first colon means how many rows we want
#to include from the imported dataset csv file.
#In this case no number specified means we want to
#include all rows.
#Right of the second colon the number specified
#indicate how many columns we want to include. 
#in this case all columns except the last one.
#.values means we want to get all the values in 
#the csv file. We assign all these values to the
#X variable as a matrix (or array of arrays) data type.
X = dataset.iloc[:, :-1].values

# Create the Dependant Variable Vector (dependant variables):
# In this case we just need to include the last column in our
#imported dataset. We start counting from column Country, starting
#with the count of 0 (with python you always count from index 0).
#So in this case column Purchased is column index 3. We don't include
#the second colon as we only want to include one column from the end.
#Not all column with an exception of the column at the end.
y = dataset.iloc[:, 3].values


# Taking care of missing data - calculate mean
# ============================================
# sklearn (or sikat learn) is a framework that contains amazing libraries
# to make machine learning models. From sklearn we import 'preprocessing'
# library that contains a lot of classes and methods to preprocess
# any data sets. We're then importing the Imputer class which takes
#care of those missing data.
from sklearn.preprocessing import Imputer
# NaN is what is represented when there's missing data in the matrix dataset.
# 'mean' is actually the default value for strategy param. So we normally 
# wouldn't need to include it, but we include it here just for readibility.
# axis param indicate whether we're calculating the mean of an entire column,
# or row to get the missing value. Of course in this case it's a column, so
# the value is 0 (otherwise 1 for row).
imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)
# we use 'fit' method to pass which matrix we want to calculate missing data,
# in this case X. Then we specify which columns we want to include. First colon
# without any numbers means include all rows, and 1:3, means include the second
# (python starts with index 0) and 3  seems to mean column 4, but it's actually 
# column 3, because the upper bound or column 3 (4 column in reality) is not to
# be included. So if you say 0:1 it means include column from 0 to 1, but 1 is
# not included. Same 1:3 means 1 to 3 but exclusive of 3 (meaning 1 and 2). 
# we're assigning all of this to the same object imputer, to make X matrix fit
# in to the object.
imputer = imputer.fit(X[:, 1:3])
# replace the missing data with the mean
X[:, 1:3] = imputer.transform(X[:, 1:3]) 


# Encoding categoral data - changing string data to number categories for ease of mathematical modelling
# ============================================
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
# an instance of LabelEncoder that represent the X matrix.
labelencoder_X = LabelEncoder()
# We fit the first column of matrix X (country column) to object labelencoder. 
# the fit_transform would return an encoded data of that column, i.e. in 
# numerical categories rather than string data. We then assign this to first
# column of matrix X to update the matrix with this change.
X[:, 0] = labelencoder_X.fit_transform(X[:, 0])

# Create the dummy data, as these are strings and there's no number order 
# relation between France, Germany, Spain ..etc. So category numbers should be
# dummy number data, i.e. 1 if country exist in the entry and 0 if it doesn't.
# First param we specify which column we want to One Hot Encode. In this case
# it's column 0 (country)
onehotencoder = OneHotEncoder(categorical_features = [0])
# now that column 0 is one-hot-encoded, we just fit it back to Matrix X.
X = onehotencoder.fit_transform(X).toarray()

# Encode the Dependant Variable Vector. As it is Dependant Variable Vector, the
# Machine Learning model will know automatically that it is a category and 
# that there's no order between the two data types. So we only need to 
# LabelEncode it. No need for OneHotEncode. 
labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)


# Splitting the dataset into a Training set and a Test set
# ========================================================
from sklearn.model_selection import train_test_split
# X_train is the training part of the Matrix Of Features
# X_test is the test part of the Matrix Of Features
# y_train is the training part of the Dependant Variable Vector (that is associated to X_train, that means that we have the same indices for both with the same entries)
# y_test is the test part of the Dependant Variable Vector (associate to X_test)
# 1st param is the X and y matrices. 2nd param is the size of the test dataset.
# A good size for the test data would be between 20% to 30% of the entire 
# dataset. In this case we specified 20% (0.2).
# In our case we have 10 entries (or 10 observations) in our dataset, meaning
# 2 observations will be test data and the remaining are training data.
# 3rd param is just random number state generator. If you want the same result 
# as say another data-scientist or analyst, who's also training and testing the
# the data so that you can both compare and make sure the ML model outputs 
# the same consist result given the same input data, then you can both put the
# same number. In our case we simply chose 0. 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0) 

File: data_preprocessing_template.R
# Data Preprocessing

#Importing the dataset
dataset = read.csv('Data.csv')

# Creating 'Matrix Of Features' and 'Dependant Variable Vector'.
#===============================================================
# Unlike Python, in R langauge, we don't need to make a 
# a distinction between the two matrices (i.e. Matrix Of Features' and 'Dependant Variable Vector').

# Taking care of missing data
#==============================
# the $ sign allows us to reference the column in the matrix
# ifelse() function takes 1st param which is the condition, 2nd param the value if the 
# condition is true, and 3rd param is the value if the condition is false. 
# is.na is a function that checks if the value in the passed to the function is missing or not.
# so we're checking if all the values in the column Age is missing. So it returns true
# if a value in the column is missing, otherwise false if there's no missing value. 
# ave is the average function in R. First param is dataset$Age column. Second param is
# where we create a new function where in the function body we use the mean R function
# passing x (the matrix i.e. dataset$Age) and na.rm = TRUE means include the missing 
# value(s) when calculating the mean of the column.
# Third param is otherwise if false (no missing values) then return the column as it is.
dataset$Age = ifelse(is.na(dataset$Age),
                     ave(dataset$Age, FUN = function(x) mean(x, na.rm = TRUE)),
                     dataset$Age)
# Take care of missing data for Salary column 
dataset$Salary = ifelse(is.na(dataset$Salary),
                     ave(dataset$Salary, FUN = function(x) mean(x, na.rm = TRUE)),
                     dataset$Salary)

# Encoding categoral data 
#=========================
# In R we factors to encode category data. 
# the factor takes param 1 as the column, param 2 as all the values in the column,
# and param 3, the encode label for each value respectively. 
# Unlike Python, R does not associate 1,2,3 as some sort of order data. So Machine Learning
# models can treat them as encoded categories and not ordered data of ascending order.
dataset$Country = factor(dataset$Country, 
                         levels = c('France', 'Spain', 'Germany'),
                         labels = c(1, 2, 3))
dataset$Purchased = factor(dataset$Purchased, 
                         levels = c('No', 'Yes'),
                         labels = c(0, 1))

# Splitting the dataset into the Training set and Test set
#===========================================================
# the line below is how to install a new package in R that is not included in the existing
# list of packages. One installed, you can comment out the line as we won't need to 
# exectue it again.
# install.packages('caTools')
# the line below is how to include a library in your code. There's 2 ways. Either tick it
# from the list of packages in the Packages tab. Or you can use the line below instead.
library(caTools)
# the line below is the same as the random_set function in python.
set.seed(123)
# splitting the data into training and test set:
# Unlike in python, in R we only need to input the Y (or dependant variable) matrix to
# the split function.
# Unlike python, here in R the SplitRatio of dataset is aimed at the Training set (not the test set). 
# So in this case we specify 80% (0.8). 
# The return value will be either true or false for each entry (or observation). 
# True if the observation is chosen to go to the Training set and false if it's chosen
# for the Test set
split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# create the training set. As the split is already created with TRUE for training set,
# and FALSE for test set, we simply assign these 2 datasets to variable as per below.
training_set = subset(dataset, split == TRUE)
# create the test set
test_set = subset(dataset, split == FALSE)


# Feature Scaling 
# Note Country and Purchased columns are factored in R, which is not the same as
# one-hot-encoding applied to categorical data in python. Even though in the 
# dataset it shows numbers in those columns, but in fact a Factor in R
# is not a numerical data. Therefore to scale the dataset directly in R in our case
# will throw an error, because not all data are numerical. In this case we need
# to exclude categories from the feature scaling. Therefore we simply include 
# the columns with the numerical values (i.e. Age and Salary, and exclued Country and Purchased).
# We do this simply by specifying training_set[, 2:3], which is column 2 and 3. 
# Remember R index count start from 1 (unlike python which starts at 0). Same applies
# to the test_set.
training_set[, 2:3] = scale(training_set[, 2:3])
test_set[, 2:3] = scale(test_set[, 2:3]) 